util/topgen/lib.py

```diff
@@ -264,23 +265,32 @@ class Name:
     internal representation into "ExampleName".
     """
     def __add__(self, other):
-        return Name(self.parts + other.parts)
+        return Name(self._parts + other._parts)
+
+    def __repr__(self):
+        return "Name({})".format(self._parts)
+
+    def __hash__(self):
+        return hash(tuple(self._parts))
+
+    def __eq__(self, other):
+        return self._parts == other._parts
 
     @staticmethod
     def from_snake_case(input: str) -> 'Name':
         return Name(input.split("_"))
 
     def __init__(self, parts: List[str]):
-        self.parts = parts
+        self._parts = parts
```

Comment:
I'd suggest converting parts to a tuple here as the typing suggests a list can be passed. Which means technically `parts` and now `_parts` is just a reference to a mutable list object... which can still be mutated outside this class object. If you want to guarantee that it won't change then you need to convert it to an immutable container like tuple at the start of the lifetime of the `Name` object.

This means you can remove the tuple conversion from the `__hash__` method as it's already done upstream. As it stands the hash could still change every time it's called if the list is mutated between calls. Which could lead to some nasty bugs.